### ISSUE TITLE
setTerrain causes hillshading opacity to increase and detail to decrease

### ISSUE THREAD

**Issue by zdila** [maplibre/maplibre-gl-js#2035]
**maplibre-gl-js version**: 2.4.0 or 3.0.0-pre.3

**browser**: Vivaldi, Chrome, Firefox

### Steps to Trigger Behavior

 1. render map with hillshading
 2. enable 3D terrain

### Link to Demonstration

https://codesandbox.io/s/maplibre-gl-js-stron-shading-issue-30pke4

### Expected Behavior

After pressing "Set terrain" button shading opacity should not change and hillshading detail should stay the same.

### Actual Behavior

After pressing "Set terrain" button the shading opacity rapidly increases and hillshading detail decreases.

**Comment by HarelM** [comment]
Is this the same as #1317?

**Comment by HarelM** [comment]
Probably not, this is strange...
Can you play around with the style to create a more simple reproduction?
I haven't seen this for other hillshade stuff, so it might be some configuration of the hillshare layer.
If possible, please create a reproduction that the entire style is in the codesandbox so that we'll be able to narrow down what's the cause of this issue.
Thanks!

**Comment by acalcutt** [comment]
One thing that I notice is you are using the same 'terrain_rgb' source for both terrain and hillshade. In the past this was not recommended due to other issues like this (i'd have to look back and find the PR)... I'm not sure if that has changed.

I the demo style for https://demotiles.maplibre.org/terrain-tiles/ it uses the same source twice.
https://demotiles.maplibre.org/styles/osm-bright-gl-terrain/style.json
```
	"sources": {
		"openmaptiles": {
			"type": "vector",
			"url": "https://demotiles.maplibre.org/tiles-omt/tiles.json"
		},
		"TerrainSource": {
			"type": "raster-dem",
			"url": "https://demotiles.maplibre.org/terrain-tiles/tiles.json"
		},
		"HillshadeSource": {
			"type": "raster-dem",
			"url": "https://demotiles.maplibre.org/terrain-tiles/tiles.json"
		}
	},
```

**Comment by acalcutt** [comment]
I think this is what I was thinking of https://github.com/maplibre/maplibre-gl-js/issues/1186#issuecomment-1100927180

**Comment by rotu** [comment]
There are some warnings picked up by webgl-lint around terrain. Possibly related, possibly a red herring. https://osmus.slack.com/archives/C01G3D28DAB/p1671217859071759

**Comment by acalcutt** [comment]
I added a second source for terrain to you example here 
https://codesandbox.io/s/maplibre-gl-js-stron-shading-issue-addsource-q15nth

Seems to look ok in that example.

**Comment by HarelM** [comment]
Assigned bounty L.
Link to parent Bounty: https://github.com/maplibre/maplibre/issues/189

**Comment by prozessor13** [comment]
I think it is best practise to use 2 sources. This is needed, because the SoureCache for terrain and hillshading have different requirements (at least the tilesize, as i remember).

* Terrain has to be fast (overscale 256x256 tiles to 1024x1024 tiles if i remember right). Also ZL 12 is sufficient.
* Hillshading needs a lot of details and needs at least ZL 14 Tiles

Sure you can refacture SourceCache logic to reuse one source for different SourceCaches, but currently this is 1:1 relation. In my opinion it would be best to simply write a good documentation for this.

**Comment by HarelM** [comment]
Good documentation is always good, but I think there's also a need to send an error or warning to the console to let the user know this is not "a great idea".

**Comment by HarelM** [comment]
Reducing bounty to M.
Expected solution:
1. Improve terrain documentation and inform the reader that a designed source for terrain is recommended
2. Add warnonce in case the source is used for other layers.

If 2 is complicated we can split the first part to a S bounty and leave the rest to a different issue.

**Comment by tempranova** [comment]
Hey there! Would be happy to take this one.

**Comment by tempranova** [comment]
Hey there! Have added some basic fixes for this:

https://github.com/maplibre/maplibre-gl-js/pull/2298 - code fix

https://github.com/maplibre/maplibre-gl-js-docs/pull/440 - doc fix

Happy to review or change verbiage or the approach if there is a better way to resolve these. I only solved for the basic case given here (ie, the warnonce when setTerrain is called), but this is an issue that a user could experience in other ways -- ie, when adding a hillshade layer after loading, or when changing visibility of hillshade layers after map load.

**Comment by HarelM** [comment]
Solved by #2298. Thanks @tempranova!!
